File: src/NodeParser/OptionalTypeNodeParser.ts
import ts from "typescript";
import type { Context, NodeParser } from "../NodeParser.js";
import type { SubNodeParser } from "../SubNodeParser.js";
import type { BaseType } from "../Type/BaseType.js";
import { OptionalType } from "../Type/OptionalType.js";

export class OptionalTypeNodeParser implements SubNodeParser {
    public constructor(protected childNodeParser: NodeParser) {}
    public supportsNode(node: ts.OptionalTypeNode): boolean {
        return node.kind === ts.SyntaxKind.OptionalType;
    }
    public createType(node: ts.OptionalTypeNode, context: Context): BaseType {
        const type = this.childNodeParser.createType(node.type, context);
        return new OptionalType(type);
    }
}

File: internal/parser/optional_type_node_parser.go
package parser

import (
	"github.com/microsoft/typescript-go/shim/ast"

	"github.com/vega/ts-json-schema-generator/internal/types"
)

// OptionalTypeNodeParser handles optional tuple elements like `[string?]`
// (src/NodeParser/OptionalTypeNodeParser.ts).
type OptionalTypeNodeParser struct {
	childNodeParser NodeParser
}

func NewOptionalTypeNodeParser(childNodeParser NodeParser) *OptionalTypeNodeParser {
	return &OptionalTypeNodeParser{childNodeParser: childNodeParser}
}

func (p *OptionalTypeNodeParser) SupportsNode(node *ast.Node) bool {
	return node.Kind == ast.KindOptionalType
}

func (p *OptionalTypeNodeParser) CreateType(node *ast.Node, ctx *Context, _ *types.ReferenceType) types.Type {
	return &types.OptionalType{Type: p.childNodeParser.CreateType(node.AsOptionalTypeNode().Type, ctx, nil)}
}
